xdp-loader/xdp-loader.c

```diff
@@ -370,6 +371,62 @@ int do_clean(const void *cfg, __unused const char *pin_root_path)
 	return libxdp_clean_references(opt->iface.ifindex);
 }
 
+static const struct featuresopt {
+	struct iface iface;
+} defaults_features = {};
+
+static struct prog_option features_options[] = {
+	DEFINE_OPTION("dev", OPT_IFNAME, struct featuresopt, iface,
+		      .positional = true,
+		      .metavar = "[ifname]",
+		      .required = true,
+		      .help = "Show XDP features for device [ifname]"),
+	END_OPTIONS
+};
+
+#define CHECK_XDP_FEATURE(f) (opts.feature_flags & (f) ? "yes" : "no")
+
+static int iface_print_xdp_features(const struct iface *iface)
+{
+#ifdef HAVE_LIBBPF_BPF_XDP_QUERY
+	LIBBPF_OPTS(bpf_xdp_query_opts, opts);
+	int err;
+
+	err = bpf_xdp_query(iface->ifindex, 0, &opts);
+	if (err)
+		return err;
+
+	/* NETDEV_XDP features are defined in <linux/netdev.h> kernel header */
+	printf("NETDEV_XDP_ACT_BASIC:\t\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_BASIC));
+	printf("NETDEV_XDP_ACT_REDIRECT:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_REDIRECT));
+	printf("NETDEV_XDP_ACT_NDO_XMIT:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_NDO_XMIT));
+	printf("NETDEV_XDP_ACT_XSK_ZEROCOPY:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_XSK_ZEROCOPY));
+	printf("NETDEV_XDP_ACT_HW_OFFLOAD:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_HW_OFFLOAD));
+	printf("NETDEV_XDP_ACT_RX_SG:\t\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_RX_SG));
+	printf("NETDEV_XDP_ACT_NDO_XMIT_SG:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_NDO_XMIT_SG));
+
+	return 0;
+#else
+	__unused const void *i = iface;
+
+	return -EOPNOTSUPP;
```

Comment:
We should print a friendly error message here, not just return an error code. Same thing when the libbpf feature is there, but the kernel doesn't support querying features...